readme.txt
=== WooCommerce Product Bundles ===

Contributors: SomewhereWarm
Tags: woocommerce, product, bundle, bundles, kits, simple, variable, configurable
Requires at least: 4.4
Tested up to: 4.9
Stable tag: 5.7.10
WC requires at least: 3.0
WC tested up to: 3.4
License: GNU General Public License v3.0
License URI: http://www.gnu.org/licenses/gpl-3.0.html

Offer product bundles and assembled products in your WooCommerce store.

== Description ==

Create assembled products and offer product bundles by grouping simple, variable and subscription products.

Looking for help? Read the full documentation [here](http://docs.woocommerce.com/document/bundles/).


changelog.txt
*** WooCommerce Product Bundles ***

2018.05.24 - version 5.7.10
* Tweak - Added a section in the WordPress Privacy Policy Guide to let store owners know that Product Bundles does not collect, store or share personal data.

2018.05.10 - version 5.7.9
* Tweak - Fixed incorrect use of 'woocommerce' text domain in some strings. Sorry about that.
* Tweak - Added back extension template overrides in WooCommerce System Report.
* Tweak - Improved integration with Storefront 2.3.
* Fix - Showing/hiding list of bundled products with Insufficient Stock in 'product' post type admin screen.
* Fix - Prevent remaining bundled product stock quantity from displaying if needed to respect the global WooCommerce stock display settings.
* Fix - Make it possible to define custom bundled item titles when calling 'WC_PB_Cart::add_bundle_to_cart' if 'Override Title' is enabled.

2018.04.20 - version 5.7.8
* Dev - Added 'woocommerce_bundled_item_discounted_price_precision' filter. Use it to set a custom precision for calculating discounted bundled product (per-unit) prices.
* Fix - WooCommerce Subscriptions: Paying for failed initial/parent order with bundled subscription products creates new order and subscription.
* Fix - "Now" part missing from catalog price of bundles containing a combination of subscription and non-subscription products.
* Fix - PHP warning when a Bundle with an infinite max price is included in a Composite Product component that's Priced Individually. Affects sites running PHP 7.1+.
* Tweak - Bundled cart line item subtotals are now aggregated as displayed.
* Tweak - Speed up removing/restoring bundles in the cart by using 'woocommerce_remove_cart_item' and 'woocommerce_restore_cart_item' instead of 'woocommerce_cart_item_removed' and 'woocommerce_cart_item_restored'.

2018.03.19 - version 5.7.7
* Fix - Prices of bundled variations loaded via AJAX are visible although 'Priced Individually' is disabled.
* Fix - Unable to select 'Group Mode > No parent' when a bundle contains virtual products.
* Fix - Support 'Custom Price Field' inputs in WooCommerce Product Add-Ons.
* Fix - PHP warning triggered when a bundle contains Subscriptions with a combined recurring price equal to zero.
* Tweak - Improved compatibility with 'WooCommerce Print Invoices and Packing Lists'.

2018.02.19 - version 5.7.6
* Fix - Fixed out-of-base client-side totals calculation in single-product pages when prices are defined including tax.
* Dev - Added 'woocommerce_bundles_compatibility_modules' filter for disabling compatibility modules.

2018.02.11 - version 5.7.5
* Fix - Bundled item runtime cache misses. Resolves performance issues when loading large bundles. Issue introduced in version 5.7.4.

2018.02.08 - version 5.7.4
* Fix - Wrong out-of-base catalog prices of Bundles with individually-priced items when prices defined incl. tax.
* Fix - When aggregating bundled item weights into the container item, bundled variation weights are ignored when inherited by their parent variable product.
* Tweak - Refactored bundle validation logic to make it easier to extend. Modified most back-end validation strings along the way -- these are displayed under very rare circumstances.
* Tweak - Hide "Shipped Individually" option for virtual bundled products.

2018.01.26 - version 5.7.3
* Fix - PHP warnings when updating 'array' type properties using v2 of the WC REST API.
* Tweak - Changed default bundled item image size to 'woocommerce_thumbnail' for consistency with WC 3.3 image size changes (WC 3.3+ only).

2018.01.09 - version 5.7.2
* Fix - Improved presentation of "Group mode > No parent" bundles in Composite Products.
* Fix - Incorrect calculation of totals affecting bundles with variable products and a single active variation with overriden attributes.
* Fix - Added support for Product Add-Ons 2.9.1.

2017.12.15 - version 5.7.1
* Fix - Contents of assembled bundles not visible in Shipstation.
* Fix - Issues with DB updating and stock status syncing of bundles & bundled products in a WP MultiSite environment - see https://github.com/woocommerce/woocommerce/pull/18060
* Fix - Hide bundled cart item price when parent is part of a Composite.

2017.12.04 - version 5.7.0
* Feature - New 'Form Location' option, located under 'Product Data > Advanced'. Allows you to modify the display location of the add-to-cart form in single-product pages. "Default": The add-to-cart form is displayed inside the single-product summary. "After summary": The add-to-cart form is displayed after the single-product summary. "After summary" usually allocates the entire page width for displaying form content. Note that some themes may not support this option.
* Feature - Client-side validation of bundled cart item quantities.
* Fix - Incorrect validation message when attempting to add a bundle with "Sold Individually" items to the cart and the quantity of the bundle is higher than 1. Now the quantity of the "Sold Individually" item is correctly kept equal to 1 in the cart.

2017.11.27 - version 5.6.2
* Fix - Prevent importing data when a CSV column is omitted.
* Fix - Totals in single-product page include the prices of sold-individually bundled items, even when optional and not selected.
* Fix - WooCommerce Min/Max Quantities integration fixes.

2017.11.20 - version 5.6.1
* Fix - '0' bundled product quantities in cart when WooCommerce Min/Max Quantities extension is installed.

2017.11.14 - version 5.6.0
* Important - Updated 'bundled-variation.php' template.
* Feature - New "Product Bundles - Bundle-Sells" mini-extension. Grab it at https://github.com/somewherewarm/woocommerce-product-bundles-bundle-sells.
* Feature - New "Product Bundles - Bulk Discounts" mini-extension. Grab it at https://github.com/somewherewarm/woocommerce-product-bundles-bulk-discounts.
* Fix - Repositioned variable bundled product add-ons under variation details.

woocommerce-product-bundles.php
<?php
/*
* Plugin Name: WooCommerce Product Bundles
* Plugin URI: https://woocommerce.com/products/product-bundles/
* Description: Offer product bundles and assembled products in your WooCommerce store.
* Version: 5.7.10
* Author: SomewhereWarm
* Author URI: https://somewherewarm.gr/
*
* Woo: 18716:fbca839929aaddc78797a5b511c14da9
*
* Text Domain: woocommerce-product-bundles
* Domain Path: /languages/
*
* Requires at least: 4.4
* Tested up to: 4.9
*
* WC requires at least: 3.0
* WC tested up to: 3.4
*
* Copyright: © 2017-2018 SomewhereWarm SMPC.
* License: GNU General Public License v3.0
* License URI: http://www.gnu.org/licenses/gpl-3.0.html
*/

// Exit if accessed directly.
if ( ! defined( 'ABSPATH' ) ) {
	exit;
}

/*
 * Required functions.
 */
if ( ! function_exists( 'woothemes_queue_update' ) ) {
	require_once( 'woo-includes/woo-functions.php' );
}

/*
 * Plugin updates.
 */
woothemes_queue_update( plugin_basename( __FILE__ ), 'fbca839929aaddc78797a5b511c14da9', '18716' );

/*
 * WC active check.
 */
if ( ! is_woocommerce_active() ) {
	return;
}

/**
 * Main plugin class.
 *
 * @class    WC_Bundles
 * @version  5.7.10
 */
class WC_Bundles {

	public $version  = '5.7.10';
	public $required = '3.0.0';

	/**
	 * The single instance of the class.
	 * @var WC_Bundles
	 *
	 * @since 4.11.4
	 */
	protected static $_instance = null;

	/**
	 * Main WC_Bundles instance. Ensures only one instance of WC_Bundles is loaded or can be loaded - @see 'WC_PB()'.
	 *
	 * @static
	 * @return WC_Bundles
	 * @since  4.11.4
	 */
	public static function instance() {
		if ( is_null( self::$_instance ) ) {
			self::$_instance = new self();
		}
		return self::$_instance;
	}

	/**
	 * Cloning is forbidden.
	 *
	 * @since 4.11.4
	 */
	public function __clone() {
		_doing_it_wrong( __FUNCTION__, __( 'Foul!', 'woocommerce-product-bundles' ), '4.11.4' );
	}

	/**
	 * Unserializing instances of this class is forbidden.
	 *
	 * @since 4.11.4
	 */
	public function __wakeup() {
		_doing_it_wrong( __FUNCTION__, __( 'Foul!', 'woocommerce-product-bundles' ), '4.11.4' );
	}

	/**
	 * Make stuff.
	 */
	protected function __construct() {
		// Entry point.
		add_action( 'plugins_loaded', array( $this, 'initialize_plugin' ), 9 );
	}

	/**
	 * Auto-load in-accessible properties.
	 *
	 * @param  mixed  $key
	 * @return mixed
	 */
	public function __get( $key ) {
		if ( in_array( $key, array( 'compatibility', 'cart', 'order', 'display' ) ) ) {
			$classname = 'WC_PB_' . ucfirst( $key );
			return call_user_func( array( $classname, 'instance' ) );
		}
	}

	/**
	 * Plugin URL getter.
	 *
	 * @return string
	 */
	public function plugin_url() {
		return untrailingslashit( plugins_url( '/', __FILE__ ) );
	}

	/**
	 * Plugin path getter.
	 *
	 * @return string
	 */
	public function plugin_path() {
		return untrailingslashit( plugin_dir_path( __FILE__ ) );
	}

	/**
	 * Plugin base path name getter.
	 *
	 * @return string
	 */
	public function plugin_basename() {
		return plugin_basename( __FILE__ );
	}

	/**
	 * Fire in the hole!
	 */
	public function initialize_plugin() {

		// WC version sanity check.
		if ( version_compare( WC()->version, $this->required ) < 0 ) {
			$notice = sprintf( __( 'WooCommerce Product Bundles requires at least WooCommerce <strong>%s</strong>.', 'woocommerce-product-bundles' ), $this->required );
			require_once( 'includes/admin/class-wc-pb-admin-notices.php' );
			WC_PB_Admin_Notices::add_notice( $notice, 'error' );
			return false;
		}

		$this->define_constants();
		$this->includes();

		WC_PB_Compatibility::instance();
		WC_PB_Cart::instance();
		WC_PB_Order::instance();
		WC_PB_Display::instance();

		// Load translations hook.
		add_action( 'init', array( $this, 'load_translation' ) );
	}

	/**
	 * Constants.
	 */
	public function define_constants() {

		wc_maybe_define_constant( 'WC_PB_SUPPORT_URL', 'https://woocommerce.com/my-account/marketplace-ticket-form/' );
		wc_maybe_define_constant( 'WC_PB_ABSPATH', trailingslashit( plugin_dir_path( __FILE__ ) ) );

		if ( 'yes' === get_option( 'woocommerce_product_bundles_debug_stock_sync', null ) || 'yes' === get_option( 'woocommerce_product_bundles_debug_stock_cache', null ) || defined( 'WC_PB_DEBUG_STOCK_CACHE' ) ) {
			/**
			 * 'WC_PB_DEBUG_STOCK_SYNC' constant.
			 *
			 * Used to disable bundled product stock meta syncing for bundled items.
			 */
			wc_maybe_define_constant( 'WC_PB_DEBUG_STOCK_SYNC', true );
		}

		if ( 'yes' === get_option( 'woocommerce_product_bundles_debug_stock_parent_sync', null ) || defined( 'WC_PB_DEBUG_STOCK_SYNC' ) ) {
			/**
			 * 'WC_PB_DEBUG_STOCK_PARENT_SYNC' constant.
			 *
			 * Used to disable stock status and visibility syncing for bundles.
			 */
			wc_maybe_define_constant( 'WC_PB_DEBUG_STOCK_PARENT_SYNC', true );
		}

		if ( 'yes' === get_option( 'woocommerce_product_bundles_debug_transients', null ) ) {
			/**
			 * 'WC_PB_DEBUG_TRANSIENTS' constant.
			 *
			 * Used to disable transients caching at plugin level.
			 */
			wc_maybe_define_constant( 'WC_PB_DEBUG_TRANSIENTS', true );
		}

		if ( 'yes' === get_option( 'woocommerce_product_bundles_debug_object_cache', null ) ) {
			/**
			 * 'WC_PB_DEBUG_OBJECT_CACHE' constant.
			 *
			 * Used to disable object caching at plugin level.
			 */
			wc_maybe_define_constant( 'WC_PB_DEBUG_OBJECT_CACHE', true );
		}

		if ( 'yes' === get_option( 'woocommerce_product_bundles_debug_runtime_cache', null ) ) {
			/**
			 * 'WC_PB_DEBUG_RUNTIME_CACHE' constant.
			 *
			 * Used to disable runtime object caching at plugin level.
			 */
			wc_maybe_define_constant( 'WC_PB_DEBUG_RUNTIME_CACHE', true );
		}
	}

	/**
	 * Includes.
	 */
	public function includes() {

		// Extensions compatibility functions and hooks.
		require_once( 'includes/compatibility/class-wc-pb-compatibility.php' );

		// Data classes.
		require_once( 'includes/data/class-wc-pb-data.php' );

		// Install.
		require_once( 'includes/class-wc-pb-install.php' );

		// Functions (incl deprecated).
		require_once( 'includes/wc-pb-functions.php' );
		require_once( 'includes/wc-pb-deprecated-functions.php' );

		// Helper functions and hooks.
		require_once( 'includes/class-wc-pb-helpers.php' );

		// Data syncing between products and bundled items.
		require_once( 'includes/class-wc-pb-db-sync.php' );

		// Product price filters and price-related functions.
		require_once( 'includes/class-wc-pb-product-prices.php' );

		// Bundled Item class.
		require_once( 'includes/class-wc-bundled-item.php' );

		// Product Bundle class.
		require_once( 'includes/class-wc-product-bundle.php' );

		// Stock mgr class.
		require_once( 'includes/class-wc-pb-stock-manager.php' );

		// Cart-related bundle functions and hooks.
		require_once( 'includes/class-wc-pb-cart.php' );

		// Order-related bundle functions and hooks.
		require_once( 'includes/class-wc-pb-order.php' );

		// Front-end filters and templates.
		require_once( 'includes/class-wc-pb-display.php' );

		// Front-end AJAX handlers.
		require_once( 'includes/class-wc-pb-ajax.php' );

		// REST API hooks.
		require_once( 'includes/class-wc-pb-rest-api.php' );

		// Admin includes.
		if ( is_admin() ) {
			$this->admin_includes();
		}

		// WP-CLI includes.
		if ( defined( 'WP_CLI' ) && WP_CLI ) {
			require_once( 'includes/class-wc-pb-cli.php' );
		}
	}

	/**
	 * Admin & AJAX functions and hooks.
	 */
	public function admin_includes() {

		// Admin notices handling.
		require_once( 'includes/admin/class-wc-pb-admin-notices.php' );

		// Admin functions and hooks.
		require_once( 'includes/admin/class-wc-pb-admin.php' );
	}

	/**
	 * Load textdomain.
	 */
	public function load_translation() {
		load_plugin_textdomain( 'woocommerce-product-bundles', false, dirname( plugin_basename( __FILE__ ) ) . '/languages/' );
	}

	/*
	|--------------------------------------------------------------------------
	| Deprecated methods.
	|--------------------------------------------------------------------------
	*/

	public function woo_bundles_plugin_url() {
		_deprecated_function( __METHOD__ . '()', '5.0.0', __CLASS__ . '::plugin_url()' );
		return $this->plugin_url();
	}
	public function woo_bundles_plugin_path() {
		_deprecated_function( __METHOD__ . '()', '5.0.0', __CLASS__ . '::plugin_path()' );
		return $this->plugin_path();
	}
}

/**
 * Returns the main instance of WC_Bundles to prevent the need to use globals.
 *
 * @since  4.11.4
 * @return WC_Bundles
 */
function WC_PB() {
  return WC_Bundles::instance();
}

$GLOBALS[ 'woocommerce_bundles' ] = WC_PB();
